#!/usr/bin/env python3
"""
Test gRPC corrig√© pour le service de traduction Meeshy
Compatible avec les d√©finitions protobuf actuelles
"""

import grpc
import time
import sys
import translation_pb2
import translation_pb2_grpc

def test_grpc_service():
    """Test complet du service gRPC de traduction"""
    print("üîç Test du service gRPC de traduction Meeshy (CORRIG√â)")
    print("=" * 60)
    
    try:
        # Connexion au serveur
        print("üì° Connexion au serveur gRPC (localhost:50051)...")
        channel = grpc.insecure_channel('localhost:50051')
        stub = translation_pb2_grpc.TranslationServiceStub(channel)
        
        # Test 1: Statistiques du service
        print("\nüî∏ Test 1: Statistiques du service")
        try:
            stats_request = translation_pb2.ServiceStatsRequest()
            stats_response = stub.GetServiceStats(stats_request, timeout=5)
            print(f"   ‚úÖ Service pr√™t: {stats_response.service_ready}")
            print(f"   ‚úÖ Mod√®les charg√©s: {stats_response.models_loaded_count}")
            print(f"   ‚úÖ Langues support√©es: {stats_response.supported_languages_count}")
            print(f"   ‚úÖ Taille cache: {stats_response.cache_size}")
            print(f"   ‚úÖ Device info: {stats_response.device_info}")
        except Exception as e:
            print(f"   ‚ùå Erreur stats: {e}")
        
        # Test 2: D√©tection de langue
        print("\nüî∏ Test 2: D√©tection de langue")
        test_texts = [
            "Hello world",
            "Bonjour le monde", 
            "Hola mundo",
            "Guten Tag"
        ]
        
        for text in test_texts:
            try:
                detect_request = translation_pb2.DetectLanguageRequest(text=text)
                detect_response = stub.DetectLanguage(detect_request, timeout=5)
                print(f"   ‚úÖ '{text}' ‚Üí {detect_response.detected_language} (conf: {detect_response.confidence_score:.2f})")
                print(f"      Complexit√©: {detect_response.complexity_score:.2f}, Mod√®le: {detect_response.recommended_model_tier}")
            except Exception as e:
                print(f"   ‚ùå Erreur d√©tection '{text}': {e}")
        
        # Test 3: Traduction simple
        print("\nüî∏ Test 3: Traduction simple")
        translations = [
            ("Hello world", "en", "fr"),
            ("Bonjour le monde", "fr", "en"),
            ("Comment allez-vous ?", "fr", "es"),
            ("How are you doing today?", "en", "de")
        ]
        
        for text, src, tgt in translations:
            try:
                translate_request = translation_pb2.TranslateRequest(
                    text=text,
                    source_language=src,
                    target_language=tgt
                )
                translate_response = stub.TranslateText(translate_request, timeout=10)
                print(f"   ‚úÖ '{text}' ({src}‚Üí{tgt}) ‚Üí '{translate_response.translated_text}'")
                print(f"      Mod√®le: {translate_response.model_tier}, Temps: {translate_response.processing_time_ms}ms, Cache: {translate_response.from_cache}")
            except Exception as e:
                print(f"   ‚ùå Erreur traduction '{text}': {e}")
        
        # Test 4: Traduction multiple
        print("\nüî∏ Test 4: Traduction multiple")
        try:
            multi_request = translation_pb2.TranslateMultipleRequest(
                text="Hello, how are you today?",
                source_language="en",
                target_languages=["fr", "es", "de"]
            )
            multi_response = stub.TranslateMultiple(multi_request, timeout=15)
            print(f"   ‚úÖ Langue d√©tect√©e: {multi_response.detected_source_language}")
            for translation in multi_response.translations:
                print(f"   ‚úÖ ‚Üí {translation.target_language}: '{translation.translated_text}'")
                print(f"      Mod√®le: {translation.model_tier}, Temps: {translation.processing_time_ms}ms")
        except Exception as e:
            print(f"   ‚ùå Erreur traduction multiple: {e}")
        
        # Test 5: Langues support√©es
        print("\nüî∏ Test 5: Langues support√©es")
        try:
            lang_request = translation_pb2.SupportedLanguagesRequest()
            lang_response = stub.GetSupportedLanguages(lang_request, timeout=5)
            print(f"   ‚úÖ {len(lang_response.languages)} langues support√©es:")
            print(f"      {', '.join(lang_response.languages)}")
        except Exception as e:
            print(f"   ‚ùå Erreur langues support√©es: {e}")
        
        print("\n‚úÖ Tests termin√©s avec succ√®s !")
        return True
        
    except grpc.RpcError as e:
        print(f"‚ùå Erreur gRPC: {e.code()}: {e.details()}")
        return False
    except Exception as e:
        print(f"‚ùå Erreur g√©n√©rale: {e}")
        return False
    finally:
        channel.close()

def wait_for_server(max_attempts=30, delay=1):
    """Attend que le serveur soit disponible"""
    print(f"‚è≥ Attente du serveur gRPC (max {max_attempts}s)...")
    
    for attempt in range(max_attempts):
        try:
            channel = grpc.insecure_channel('localhost:50051')
            stub = translation_pb2_grpc.TranslationServiceStub(channel)
            
            # Test simple de connexion
            request = translation_pb2.DetectLanguageRequest(text="test")
            response = stub.DetectLanguage(request, timeout=2)
            
            channel.close()
            print("‚úÖ Serveur disponible !")
            return True
            
        except Exception:
            if attempt < max_attempts - 1:
                print(f"   Tentative {attempt + 1}/{max_attempts}...")
                time.sleep(delay)
            else:
                print("‚ùå Serveur non disponible apr√®s timeout")
                return False
    
    return False

if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Test du service gRPC de traduction")
    parser.add_argument("--wait", action="store_true", help="Attendre que le serveur soit disponible")
    parser.add_argument("--timeout", type=int, default=30, help="Timeout d'attente (secondes)")
    
    args = parser.parse_args()
    
    if args.wait:
        if not wait_for_server(max_attempts=args.timeout):
            sys.exit(1)
    
    success = test_grpc_service()
    sys.exit(0 if success else 1)
